File: Hashwanth/Arrays & Strings/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
       vector<vector<int>> ans;
        if(nums.size()<3)
            return ans;
        set<vector<int>> sans;
            sort(nums.begin(),nums.end());
       for(int i=0;i<nums.size()-2;i++)
       {  if(i!=0 && nums[i]==nums[i-1])
           continue;
        
           int l=i+1;
           int r=nums.size()-1;
           while(l<r)
           {
               vector<int> v;
               if(nums[i]+nums[l]+nums[r]==0)
               {
                   v.push_back(nums[i]);
                   v.push_back(nums[l]);
                   v.push_back(nums[r]);
                   l++;
                   r--;
                   sans.insert(v);
               }
               else if(nums[i]+nums[l]+nums[r]<0)
                   l++;
               else
                   r--;
           }
      
       }
         for(auto x:sans)
            ans.push_back(x);
        return ans;
    }
};

File: Utkarsh mishra/Arrays and String/Questions/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> r = new ArrayList<>();

        if (nums.length < 3) {
            return r;
        }

        for (int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int j = i + 1;
            int k = nums.length - 1;
            while (j < k) {
                if (nums[i] + nums[j] + nums[k] == 0) {
                    List<Integer> l = new ArrayList<>();
                    l.add(nums[i]);
                    l.add(nums[j]);
                    l.add(nums[k]);
                   r.add(l);

                    j++;
                    k--;

                    while (j < k && nums[j] == nums[j - 1]) {
                        j++;
                    }

                    while (j < k && nums[k] == nums[k + 1]) {
                        k--;
                    }
                    
                } else if (nums[i] + nums[j] + nums[k] > 0) {
                    k--;
                } else {
                    j++;
                }
            }
        }
        return r;  
    }
}
